changelog/add_traits_isCopyable.dd

```diff
@@ -0,0 +1,5 @@
+Add new builtin `__traits(isCopyable, T)` for some type `T`
+
+Changing Phobos' `std.traits.isCopyable` to make use of this new builtin trait
```

Comment:
Do you have a suggestion for how to check for a disabled copy constructor? Should I copy the logic from `e.ident == Id.hasCopyConstructor`?

Reply:
You need to check if there is a matching copy constructor overload for your type. It should be inside `isCopyable`.
This should do if you remove the `nothrow pure @nogc` attributes.
```d
        if (ts.sym.hasCopyCtor)
        {
            // check if there is a matching overload of the copy constructor and whether it is disabled or not
            Dsymbol ctor = search_function(cast() ts.sym, Id.This);
            assert(ctor);
            scope el = new IdentifierExp(Loc.initial, Id.p); // dummy lvalue
            el.type = cast() ts;
            Expressions args;
            args.push(el);
            FuncDeclaration f = resolveFuncCall(Loc.initial, null, ctor, null, cast()ts, &args, FuncResolveFlag.quiet);
            if (!f || f.storage_class & STC.disable)
                return false;
        }
```

Reply:
Done.

Reply:
>             FuncDeclaration f = resolveFuncCall(Loc.initial, null, ctor, null, cast()ts, &args, FuncResolveFlag.quiet);

Is there now direct way of looking up the specific constructor that takes `ref typeof(this)` as single parameter?

Reply:
I guess you want to cache the result instead of trying every overload each time. I don't think it's now yet, if you want to do it it has to be stored on the TypeStruct instead of the StructDeclaration because there can multiple types with different type modifiers (cons, immutable, shared... etc) which all share the same StructDeclaration but each will match a different overload for the copy constructor.

Reply:
Can we add this caching in another pull request have this merged?

Reply:
Since it's not required, yes.